<launch>
    <!-- you can write your own calibrate.launch, by including the ROS drivers for robot and tracking and choosing
    your own namespace if you have multiple tracking devices or robots -->

    <arg name="start_simulator" default="true" doc="Start a simulated tracking system with a trivial noise model" />

    <arg name="namespace_prefix" default="demo_handeyecalibration" />
    <arg name="eye_on_hand" default="false" />
    <arg name="use_iiwa" default="false" />
    <arg name="use_interbotix" default="true" />
    <arg name="robot_name" default="px150"/>



    <arg if="$(arg use_iiwa)" name="robot_base_frame" value="iiwa_link_0" />
    <arg if="$(arg use_iiwa)" name="robot_effector_frame" value="iiwa_link_ee" />
    <arg if="$(arg use_iiwa)" name="manipulator_group_name" value="manipulator" />



    <arg if="$(arg use_interbotix)" name="robot_base_frame" value="$(arg robot_name)/base_link"/>
    <arg if="$(arg use_interbotix)" name="robot_effector_frame" value="$(arg robot_name)/ee_arm_link"/>
    <arg if="$(arg use_interbotix)" name="manipulator_group_name" value="interbotix_arm" />


    <group if="$(arg use_iiwa)">
      <include file="$(find flex_grasp)/launch/iiwa_enviroment.launch">
          <arg name="rviz" value="false" />
      </include>
    </group>


    <group if="$(arg use_interbotix)">
      <include file="$(find flex_grasp)/launch/interbotix_enviroment.launch">
          <arg name="rviz" value="false" />
      </include>
    </group>

    <arg name="tracking_base_frame" value="tracking_origin" />
    <arg name="tracking_marker_frame" value="tracking_marker" />

    <!-- start the simulated tracking -->
    <node name="tracking_simulator" pkg="easy_handeye_demo" type="tracking_simulator.py" output="screen">
        <!-- if false, it computes the tracking output with respect to the ground truth and publishes it wrt the calibration result -->
        <param name="is_calibration" value="true" />

        <!-- tracking simulation frequency and noise parameters -->
        <param name="frequency" value="10" />
        <param name="translation_noise_stdev" value="0.001" />
        <param name="rotation_noise_stdev" value="0.0001" />

        <param name="eye_on_hand" value="$(arg eye_on_hand)" />
        <!-- transformations for the eye-on-base case -->
        <param unless="$(arg eye_on_hand)" name="ground_truth_calibration_transformation" value="1 0 0.5 0 0 0 1" />
        <param unless="$(arg eye_on_hand)" name="arbitrary_marker_placement_transformation" value="0.12 0.21 0.137 0 0 0 1" />
        <!-- transformations for the eye-on-hand case -->
        <param if="$(arg eye_on_hand)" name="ground_truth_calibration_transformation" value="0.12 0.21 0.137 0 0 0 1" />
        <param if="$(arg eye_on_hand)" name="arbitrary_marker_placement_transformation" value="1 0 0.5 0 0 0 1" />

        <param name="tracking_base_frame" value="$(arg tracking_base_frame)" />
        <param name="tracking_marker_frame" value="$(arg tracking_marker_frame)" />
        <param name="robot_base_frame" value="$(arg robot_base_frame)" />
        <param name="robot_effector_frame" value="$(arg robot_effector_frame)" />

        <!-- simulated tracking noise parameters -->
        <param name="frequency" value="10" />
        <param name="translation_noise_stdev" value="0.0001" />
        <param name="rotation_noise_stdev" value="0.00001" />
    </node>

    <!-- start easy_handeye calibration procedure -->
    <include ns="/$(arg robot_name)" file="$(find easy_handeye)/launch/calibrate.launch" >
        <arg name="start_rviz" value="false" />

        <arg name="namespace_prefix" value="$(arg namespace_prefix)" />
        <arg name="eye_on_hand" value="$(arg eye_on_hand)" />
        <arg name="publish_dummy" value="false" /> <!-- in our case it is done by the tracking simulator -->

        <arg name="tracking_base_frame" value="$(arg tracking_base_frame)" />
        <arg name="tracking_marker_frame" value="$(arg tracking_marker_frame)" />
        <arg name="robot_base_frame" value="$(arg robot_base_frame)" />
        <arg name="robot_effector_frame" value="$(arg robot_effector_frame)" />

        <arg name="freehand_robot_movement" value="false" />
        <arg name="robot_velocity_scaling" value="0.5" />
        <arg name="robot_acceleration_scaling" value="0.2" />

        <arg name="move_group" value="$(arg manipulator_group_name)"/>
    </include>

    <!-- start rviz with custom configuration -->
    <!-- ns="$(arg robot_name)" -->
    <node ns="$(arg robot_name)" name="$(anon rviz)" pkg="rviz" type="rviz" respawn="true"
          args="-d $(find easy_handeye_demo)/launch/iiwa_stack_config.rviz" output="screen" />
</launch>
